migrations/2021-01-27-202728_active_users_monthly/up.sql

```diff
@@ -0,0 +1,89 @@
+-- Add monthly and half yearly active columns for site and community aggregates
+
+-- These columns don't need to be updated with a trigger, so they're saved daily via queries
+alter table site_aggregates add column users_active_day bigint not null default 0;
+alter table site_aggregates add column users_active_week bigint not null default 0;
+alter table site_aggregates add column users_active_month bigint not null default 0;
+alter table site_aggregates add column users_active_half_year bigint not null default 0;
+
+alter table community_aggregates add column users_active_day bigint not null default 0;
+alter table community_aggregates add column users_active_week bigint not null default 0;
+alter table community_aggregates add column users_active_month bigint not null default 0;
+alter table community_aggregates add column users_active_half_year bigint not null default 0;
+
+create or replace function site_aggregates_activity(i text)
+returns int
+language plpgsql
+as
+$$
+declare
+   count_ integer;
+begin
+  select count(*) 
+  into count_
+  from (
+    select c.creator_id from comment c
+    inner join user_ u on c.creator_id = u.id
+    where c.published > ('now'::timestamp - i::interval) 
+    and u.local = true
+    union
+    select p.creator_id from post p
+    inner join user_ u on p.creator_id = u.id
+    where p.published > ('now'::timestamp - i::interval)
+    and u.local = true
+  ) a;
+  return count_;
+end;
+$$;
+
+update site_aggregates 
+set users_active_day = (select * from site_aggregates_activity('1 day'));
+
+update site_aggregates 
+set users_active_week = (select * from site_aggregates_activity('1 week'));
+
+update site_aggregates 
+set users_active_month = (select * from site_aggregates_activity('1 month'));
+
+update site_aggregates 
+set users_active_half_year = (select * from site_aggregates_activity('6 months'));
+
+create or replace function community_aggregates_activity(i text)
+returns table(count_ bigint, community_id_ integer)
+language plpgsql
+as
+$$
+begin
+  return query 
+  select count(*), community_id
+  from (
+    select c.creator_id, p.community_id from comment c
+    inner join post p on c.post_id = p.id
```

Comment:
For community counts tho, I don't use a `community.local` or `user.local` filter, because federated users might be doing lots of things on your community. Or you might be viewing a non-local community on your instance, and you still want to know the active users.

Reply:
For remote communities you are not gonna see the correct user count, only the number of users from your instance which are subscribed to that community. Only if you are viewing the community on the instance where it is hosted will you see the correct follower count this way. It would be better to use the follower count from `/c/{community}/followers` (although that can be manipulated).

Reply:
> For remote communities you are not gonna see the correct user count, only the number of users from your instance which are subscribed to that community.

As soon as someone subscribes to that remote community, your instance should fetch users as they comment and post over there. So the counts shouldn't be too far off. I do wanna keep these as local DB queries rather than remote fetches, even if the numbers are a little off.

Reply:
It will fetch the users, but I dont think they get inserted into the `Followers` table. That happens only on the instance where the community is, and where the user is. You can check a production DB to make sure.

Reply:
The queries actually don't care about the followers table, they only care whether any users, subscribed or not, have made a post or comment.

Reply:
Ah I see, then it should be mostly fine. Only problem is when no local user is subscribed to the community, or when local users only have been subscribed for a short time (eg, first local user subscribed a week ago, then the one month and six month statistics would be wrong). You might hide some info in those cases, or simply put a disclaimer that data might be inaccurate, and link to the original instance.